genrl/deep/agents/dqn/base.py

```diff
@@ -0,0 +1,246 @@
+
+        This updates the epsilon (exploration rate) of the agent every timestep
+
+        Args:
+            timestep (int): Timestep of training
+        """
+        self.timestep = timestep
+        self.epsilon = self.calculate_epsilon_by_frame()
+        self.logs["epsilon"].append(self.epsilon)
+
+    def select_action(
+        self, state: np.ndarray, deterministic: bool = False
+    ) -> np.ndarray:
+        """Select action given state
+
+        Epsilon-greedy action-selection
+
+        Args:
+            state (:obj:`np.ndarray`): Current state of the environment
+            deterministic (bool): Should the policy be deterministic or stochastic
+
+        Returns:
+            action (:obj:`np.ndarray`): Action taken by the agent
+        """
+        if not deterministic:
+            if np.random.rand() < self.epsilon:
+                return np.asarray(self.env.sample())
+
+        state = torch.FloatTensor(state)
+        q_values = self.model(state).detach().numpy()
+        return np.argmax(q_values, axis=-1)
+
+    def get_q_values(self, states: torch.Tensor, actions: torch.Tensor) -> torch.Tensor:
+        """Get Q values corresponding to specific states and actions
+
+        Args:
+            states (:obj:`torch.Tensor`): States for which Q-values need to be found
+            actions (:obj:`torch.Tensor`): Actions taken at respective states
+
+        Returns:
+            q_values (:obj:`torch.Tensor`): Q values for the given states and actions
+        """
+        q_values = self.model(states).gather(1, actions)
+        return q_values
+
+    def get_target_q_values(
+        self, next_states: torch.Tensor, rewards: List[float], dones: List[bool]
+    ) -> torch.Tensor:
+        """Get target Q values for the DQN
+
+        Args:
+            next_states (:obj:`torch.Tensor`): Next states for which target Q-values
+                need to be found
+            rewards (:obj:`list`): Rewards at each timestep for each environment
+            dones (:obj:`list`): Game over status for each environment
+
+        Returns:
+            target_q_values (:obj:`torch.Tensor`): Target Q values for the DQN
+        """
+        next_q_target_values = self.target_model(next_states)
+        max_next_q_target_values = next_q_target_values.max(1)[0]
+
+        target_q_values = rewards + self.gamma * torch.mul(
+            max_next_q_target_values, (1 - dones)
+        )
+        return target_q_values.unsqueeze(-1)
+
+    def get_q_loss(self) -> torch.Tensor:
+        """Function to calculate the loss of the Q-function
+
+        Returns:
+            loss (:obj:`torch.Tensor`): Calculateed loss of the Q-function
+        """
+        states, actions, rewards, next_states, dones = self.replay_buffer.sample(
+            self.batch_size
+        )
+        states = states.reshape(-1, *self.env.obs_shape)
+        actions = actions.reshape(-1, *self.env.action_shape).long()
```

Comment:
If actions are discrete, won't this already be long? Is this type casting needed?

Reply:
Yeah, when we get it from the buffer, it's converted to a FloatTensor by default. The `.gather` function needs it to be a LongTensor, hence the typecast.

Reply:
Ok